org.uqbar.project.wollok.typeSystem/src/org/uqbar/project/wollok/typesystem/constraints/variables/SimpleTypeInfo.xtend

```diff
@@ -92,19 +93,23 @@ class SimpleTypeInfo extends TypeInfo {
 	}
 
 	override ConcreteTypeState addMinType(WollokType type) {
-		if (minTypes.containsKey(type)) {
-			Ready
-		} else if (!canAddMinType(type)) {
-			throw new RejectedMinTypeException(type)
-		} else {
-			minTypes.put(type, Pending)
-			Pending
-		}
+		if (minTypes.containsKey(type)) return Ready
+		
+		validateNewMinType(type)
+		
+		minTypes.put(type, Pending)
+		Pending
+	
 	}
 
-	def canAddMinType(WollokType type) {
-		if (sealed) minTypes.keySet.exists[isSuperTypeOf(type)]
-		else type.respondsToAll(messages)
+	def validateNewMinType(WollokType type) {
+		if (sealed && !minTypes.keySet.exists[isSuperTypeOf(type)]) 
+			throw new RejectedMinTypeException(type)
+		
+		validMessages.forEach[
+			if(!type.respondsTo(it)) throw new MessageNotUnderstoodException(type, it)
```

Comment:
What happens if there is more than one message that is not understood?

Reply:
:thinking:

Reply:
It works! I think that this method is called many times and sets the messages wrong one by one. Maybe we can do that more efficiently?